pkg/controllers/rollout/controller_test.go

```diff
@@ -1503,141 +1451,787 @@ func TestPickBindingsToRoll(t *testing.T) {
-	return &fleetv1beta1.PlacementPolicy{
-		PlacementType:    placementType,
-		NumberOfClusters: ptr.To(numberOfClusters),
-		Affinity: &fleetv1beta1.Affinity{
-			ClusterAffinity: &fleetv1beta1.ClusterAffinity{
-				RequiredDuringSchedulingIgnoredDuringExecution: &fleetv1beta1.ClusterSelector{
-					ClusterSelectorTerms: []fleetv1beta1.ClusterSelectorTerm{
-						{
-							LabelSelector: &metav1.LabelSelector{
-								MatchLabels: map[string]string{
-									"key1": "value1",
-								},
-							},
-						},
+		"test bound deleting binding - rollout blocked": {
+			allBindings: []*fleetv1beta1.ClusterResourceBinding{
+				setDeletionTimeStampForBinding(generateClusterResourceBinding(fleetv1beta1.BindingStateBound, "snapshot-1", cluster1)),
+			},
+			crp: clusterResourcePlacementForTest("test",
+				createPlacementPolicyForTest(fleetv1beta1.PickAllPlacementType, 0),
+				createPlacementRolloutStrategyForTest(fleetv1beta1.RollingUpdateRolloutStrategyType, generateDefaultRollingUpdateConfig(), nil)),
+			wantTobeUpdatedBindings:     []int{},
+			wantStaleUnselectedBindings: []int{},
+			wantNeedRoll:                false,
+			wantWaitTime:                time.Second,
+		},
+		"test policy change with MaxSurge specified, evict resources on unscheduled cluster - rollout allowed for one scheduled binding": {
+			allBindings: []*fleetv1beta1.ClusterResourceBinding{
+				setDeletionTimeStampForBinding(generateCanBeReadyClusterResourceBinding(fleetv1beta1.BindingStateUnscheduled, "snapshot-1", cluster1)),
+				generateCanBeReadyClusterResourceBinding(fleetv1beta1.BindingStateUnscheduled, "snapshot-1", cluster2),
+				generateClusterResourceBinding(fleetv1beta1.BindingStateScheduled, "snapshot-1", cluster3),
+				generateClusterResourceBinding(fleetv1beta1.BindingStateScheduled, "snapshot-1", cluster4),
+			},
+			latestResourceSnapshotName: "snapshot-1",
+			crp: clusterResourcePlacementForTest("test",
+				createPlacementPolicyForTest(fleetv1beta1.PickNPlacementType, 2),
+				createPlacementRolloutStrategyForTest(fleetv1beta1.RollingUpdateRolloutStrategyType, &fleetv1beta1.RollingUpdateConfig{
+					MaxUnavailable: &intstr.IntOrString{
+						Type:   intstr.Int,
+						IntVal: 0,
 					},
+					MaxSurge: &intstr.IntOrString{
+						Type:   intstr.Int,
+						IntVal: 1,
+					},
+					UnavailablePeriodSeconds: ptr.To(1),
+				}, nil)),
+			wantDesiredBindingsSpec: []fleetv1beta1.ResourceBindingSpec{
+				{},
+				{},
+				{
+					State:                fleetv1beta1.BindingStateBound,
+					TargetCluster:        cluster3,
+					ResourceSnapshotName: "snapshot-1",
+				},
+				{
+					State:                fleetv1beta1.BindingStateBound,
+					TargetCluster:        cluster4,
+					ResourceSnapshotName: "snapshot-1",
 				},
 			},
+			wantTobeUpdatedBindings:     []int{2}, // specified MaxSurge helps us pick one scheduled binding to rollout. we don't have any ready binding so we don't remove any binding.
```

Comment:
In theory, we pick one from 2 or 3. is it guaranteed to pick 2?

Reply:
It's guaranteed we pick 2 before 3, we create bounding candidates based on order from allBindings https://github.com/Azure/fleet/blob/main/pkg/controllers/rollout/controller.go#L380 and then we chose update bindings from bounding candidates in the same order https://github.com/Azure/fleet/blob/main/pkg/controllers/rollout/controller.go#L477